### PR TITLE
feat(typeahead): allowing search on focus and click

### LINKED ISSUE #1853: TypeAhead's dropdown is closed when clicking on the input
### Bug description

When using the TypeAhead, when the dropdown is open, clicking on the input closes it.

Even though it could be seen as a feature, this is not necessarily intuitive nor wanted. If this behavior is considered acceptable and wanted (as default or not) we should at least provide an alternative where the dropdown doesn't close when clicking on the input.

For reference, at the time of writing those websites keep the dropdown open: 

- [Google](https://www.google.fr/)
- [Booking.com](https://www.booking.com/)
- [Amazon.com](https://www.amazon.com/)
- ...

### Link to minimally-working plunker that reproduces the issue

I just used the first one from the [documentation page](https://ng-bootstrap.github.io/#/components/typeahead/examples): https://ng-bootstrap.github.io/app/components/typeahead/demos/basic/plnkr.html

Here is the frozen plunkr: [Example usage of the typeahead widget from https://ng-bootstrap.github.io - Plunker](http://embed.plnkr.co/yQEbXluwIjbgfaeMMKbw/).

### Version of Angular, ng-bootstrap, and Bootstrap

- Angular: 4.0.3
- ng-bootstrap: 1.0.0-beta.4
- Bootstrap: 4.0.0-beta

### PR DESCRIPTION
**Pull request by ymeine**
Related to #698, fixes #1853

A previous PR was open (#1895) but after discussing with @pkozlowski-opensource and @maxokorokov we went for as simpler solution consisting mostly of the bug fixes and a sample demo.

### CODE REVIEW

File: src/typeahead/typeahead.ts [hunk 1]
```diff
@@ -80,7 +80,8 @@ let nextWindowId = 0;
     '[attr.aria-owns]': 'isPopupOpen() ? popupId : null',
     '[attr.aria-expanded]': 'isPopupOpen()'
   },
-  providers: [NGB_TYPEAHEAD_VALUE_ACCESSOR]
+  providers: [NGB_TYPEAHEAD_VALUE_ACCESSOR],
+  exportAs: 'ngbTypeahead'
```
Comment (maxokorokov) [inline src/typeahead/typeahead.ts]: @ymeine, maybe just move this to the top near the selector
@pkozlowski-opensource you had some concerns about exporting typeahead or I'm mistaken?

File: src/typeahead/typeahead.ts [hunk 2]
```diff
@@ -213,6 +214,21 @@ export class NgbTypeahead implements ControlValueAccessor,
...
 
+  onDocumentClick(event) {
+    if (this.isPopupOpen()) {
+      if (!this._isClickInWidget(event)) {
+        this.dismissPopup();
+      }
+    }
+  }
+
+  _isClickInWidget(event) {
```
Comment (maxokorokov) [inline src/typeahead/typeahead.ts]: should be private

File: src/util/util.ts [hunk 3]
```diff
@@ -37,3 +37,15 @@ export function padNumber(value: number) {
 export function regExpEscape(text) {
   return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
 }
+
+/**
+ * @param possibleAncestor Element to verify as being the reference element or one of its ancestors
+ * @param reference The reference element to check against
+ */
+export function isElementOrAncestorOf(possibleAncestor, reference) {
```
Comment (pkozlowski-opensource) [inline src/util/util.ts]: Couldn't we use https://developer.mozilla.org/en-US/docs/Web/API/Node/contains instead?

**Comment by pkozlowski-opensource** [comment]
@ymeine could you please rebase and check my comment regarding the `isElementOrAncestorOf`?

Another thing: it seems like this PR tackles 2 separate issues:
* clicks inside typeahead popover
* opening popup on search.

We should _really_ try to have _very minimal_ PRs so those are easier to review / track / revert if needed. Would you mind splitting this PR into 2 separate PRs, each one tackling one issue?

File: src/util/util.ts (hunk 3 above)
Comment (ymeine) [reply src/util/util.ts]: Sure, I discovered a bit later that this was used in the rest of the code. However as discussed offline indeed we don't even need to use such a feature for specific case: I can just compare the event target with the input element on which the directive is applied. For now, there is no possibility of hierarchy.

**Comment by ymeine** [comment]
@pkozlowski-opensource Comment checked and I am going to open two distinct PRs as requested, this makes sense. However, the "open on focus" depends on the "don't close on input click", so future possible single reverts is compromised.

**Comment by pkozlowski-opensource** [comment]
Closing in favour of #1989 and #1990

**Pull request closed without merging**